### ISSUE TITLE
Make mgd77list always successful

### ISSUE THREAD

**Issue by sarzeaud** [GenericMappingTools/gmt#4539]
Currently, mgd77list returns with an error when the user requests a field that is not in the data. For example:
<pre><code>mgd77list [ERROR]: Requested column az not in data set!
mgd77list [WARNING]: One or more requested columns not present in cruise 9701007062.mgd77 - skipping</code></pre>
What is skipped here is not the field, but the entire cruise.

When mgd77list is to be used in a script or another program, this behavior requires a lot of work to check what it is possible to request, and/or to get the right field in the right column. It would be much easier if mgd77list always returns a table with as many columns as requested, including NaN columns for missing fields.

When getting the result back through a VirtualFile in the API, the table is actually written in the text field of each line. It would also be more comfortable to have all the numerical fields in the data, and only the non numerical ones in the text field.

I am ready to help implement, test and maintain this feature.

Olivier

**Comment by PaulWessel** [comment]
Thanks for the offer to help, @sarzeaud.  I think the first step is to separate out a check for _incorrect_ columns before tackling _missing_ columns.  E.g., the column "asdfoiasudf" simply does not exist and should result in an error, while "bcc" should pass that test as a recognized valid column name in MGD77, even if it may possibly be _missing_ from any specific cruise file.  Right now both types of "bad" columns are treated the same.

Perhaps you could consider writing a PR that adds the new function

`int MGD77_Verify_Columns (struct GMT_CTRL *GMT, char *arg);`

which may borrow some logic from _MGD77_Select_Columns_ to extract the listed columns and then compare each such column name to the full list of recognized names (i.e., the list currently printed by the **mgd77list** usage message).  Perhaps add a static internal `char ** valid_colnames[] = {"ataime", "azim", ...}` in mgd77.c that is alphabetically sorted and let _MGD77_Verify_Columns_ check each candidate column name via _bsearch_ or similar to see if it is allowed.  Print an error message for each unrecognized columns and return a nonzero code, else return 0.  Then we would add this in **mgd77list** at the end of the parse function:

`n_errors += gmt_M_check_condition (GMT, Ctrl->F.active && MGD77_Verify_Columns (GMT, Ctrl->F.flags), "Option F: Invalid column names encountered\n");`

Reasonable?

**Comment by PaulWessel** [comment]
Thanks for #4555, I will consider the next step; busy week so may take a few days unfortunately.

**Comment by PaulWessel** [comment]
OK, still busy but have some ideas on how this can be completed.  As you know, _MGD77_Read_Header_Record_ detects that one or more columns are not present.  OK, so I suggest to make the following changes:

_MGD77_Order_Columns_: Instead of giving fatal messages and returning the error code, let's change **GMT_MSG_ERROR** to **GMT_MSG_WARNING**, set `set = GMT_NOTSET` and i`tem = i`, and remove the return call. Note: also only set `H->info[set].col[item].pos = i`; when `set != GMT_NOTSET`.  This should let us bypass the first error message.

Then, in mgd77list.c, need to look for all the instances of `M.order` where we pull out set and item, usually called _c_ and _id_ in the code.  We will need to add a new check on `c == GMT_NOTSET`.  I think not needed around line 1129 since it will go into the else branch just fine.

At line 1150 we must check if `c == GMT_NOTSET` and if so use the string` F->desired_column[id]` as column header.

Line 1566, just need to start the if test with

```
if (c = GMT_NOTSET) 
	gmt_add_to_record (GMT, record, GMT->session.d_NaN, pos, GMT_OUT, sep_flag);	/* Write NaN value */
else ...
```
 
and same around line 1601.  With some luck this may work.  Or we learn something during compilation or execution!

**Comment by sarzeaud** [comment]
On my way to make these changes, I started playing with a file of mine. I am not sure to understand what is going on here.
On this file mgd77info says:
<pre><code>
>mgd77info 9701007062.mgd77 -C
drt     tz      ptc     twt     depth   bcc     btc     mtf1    mtf2    mag     msens   diur    msdgobs    eot     faa     nqc     sln     sspn
</code></pre>
OK, then I tried to request some fields that are not listed here, namely vel and azim:
<pre><code>
>mgd77list 9701007062.mgd77 -Flon,lat,vel,azim | head
-4.50733        48.35665        4.38550827477   255.528656722
-4.51077        48.35606        4.38550827477   255.528656722
-4.51419        48.35545        4.37109880071   254.9763901
-4.51755        48.35484        4.29963791769   254.720681308
-4.52092        48.35424        4.30677062349   255.002536148
-4.52433        48.35364        4.35454484867   255.170859486
-4.52787        48.353  4.52882206474   254.781995463
-4.5316 48.35227        4.80002453059   253.591518057
-4.53549        48.35154        4.98992756408   254.232515837
-4.53941        48.35075        5.05675682312   253.13140332
</code></pre>
which looks nice, doesn't it? 
Undoubtedly, the fields are in the file, because mgd77list passes the MGD77_verify_fields function, and the values seem perfectly correct. Why does mgd77info not list them so?

I then did the modifications you suggested, but was unlucky. Compiling needed M.desired_column[id] instead of F->desired_column on line 1150 of mgd77list. But then, the output is 0 for all columns (including lon and lat) even with valid field names, and the program aborts with a segmentation fault. Not sure where to go from this point.

**Comment by PaulWessel** [comment]
No, stuff like az and vel are in no files.  They are derived on the fly from lon,lat,time.  They are what is called auxiliary parameters (including distance and a few others).  So they are not listed because not part of the files.

If you are stuck, you can make a pull request and start it with "WIP title..." and ask me to review.  When I find some time I can test it.

**Comment by sarzeaud** [comment]
I eventually caught my bug, and just made a pull request.
